### PR TITLE
box: introduce next and previous prefix iterators

### LINKED ISSUE #9994: new iterator for TREE index: `NP` (NextPrefix)
example table contains PRIMARY KEY

```yaml
- foo
- foo bar
- foo bar baz
- g
- g abc
- g cde
- g cde def
```

```lua
for _, tuple in space:pairs('foo', { iterator = 'GE' }) do
   print(tuple)
end

- foo
- foo bar
...
```

```lua
for _, tuple in space:pairs('foo', { iterator = 'NP' }) do
  print(tuple)
end

- g
- g abc

```

the iterator works like `GT` but for the first key it can receive a string that can be part of the field (prefix substring).
the first iteration must return value which prefix is more than noticed in key.

### PR DESCRIPTION
**Pull request by alyapunov**
Implement 'np' (next prefix) and 'pp' (previous prefix) iterators. They work only in memtx tree and in a nutshell searches for strings with greater ('np') or less ('pp') prefix of size as in given key, comparing with given key.

Closes #9994

@TarantoolBot document
Title:'np' and 'pp' (next/previous prefix) iterators

Now there are two more iterators avaialbe: 'np' (next prefix) and 'pp' (previous prefix). They work only in memtx tree. Also, if the last part of key is not a string, they degrade to 'gt' and 'lt' iterators.

These iterators introduce special comparison of the last part of key (if it is a string). In terms of lua, if s is the search part, and t is the corresponding tuple part, 'np' iterator searches for the first tuple with string.sub(t, 1, #s) > s, while 'pp' searches for the last tuple with string.sub(t, 1, #s) < s.

Comparison of all other parts of the key remains normal.

As usual, these iterators are available both in select and pairs, in index and space methods.

Similar to all other tree iterators, they change only initial search of selection. Once the first tuple found, the rest are selected sequentially in direct (for 'np') or reverse (for 'pp') order of the index.

For example:
```
tarantool> s:select{}
---
- - ['a']
  - ['aa']
  - ['ab']
  - ['b']
  - ['ba']
  - ['bb']
  - ['c']
  - ['ca']
  - ['cb']
...

tarantool> s:select({'b'}, {iterator = 'np'})
---
- - ['c']
  - ['ca']
  - ['cb']
...

tarantool> s:select({'b'}, {iterator = 'pp'})
---
- - ['ab']
  - ['aa']
  - ['a']
...
```

### CODE REVIEW

**Review by drewdzzz — APPROVED** [review]
LGTM

File: changelogs/unreleased/gh-9994-next-prefix-iterator.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+## feature/core
+
+* Implemented new iterators for memtx tree index: 'np' (next prefix)
```
Comment (xuniq) [inline changelogs/unreleased/gh-9994-next-prefix-iterator.md:3]:
```suggestion
* Implemented new iterators for the memtx tree index: 'np' (next prefix)
```

**Review by xuniq — APPROVED** [review]

File: test/box-luatest/gh_9994_next_prefix_iterator_test.lua [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+        s:replace{1, '/home/user/data1', 0}
+        s:replace{1, '/home/user/data2', 0}
+        s:replace{1, '/home/user/folder/', 0}
+        s:replace{1, '/home/user/folder/subfolder/data.txt', 0}
+        s:replace{1, '/home/user/folder1/subfolder/data.txt', 0}
+        s:replace{1, '/home/user/bin/tarantool', 1}
+        s:replace{1, '/home/user/bin/tarantool_ctl', 1}
+        s:replace{1, '/home/user/work/tarantool/src/main.cc', 0}
+        s:replace{1, '/home/user/work/tarantool/src/main.cc', 1}
+        s:replace{1, '/home/user/work/tarantool/src/main.cc', 0}
```
Comment (Gumix) [inline test/box-luatest/gh_9994_next_prefix_iterator_test.lua:352]: This line looks redundant.

**Review by Gumix — APPROVED** [review]

File: changelogs/unreleased/gh-9994-next-prefix-iterator.md (hunk 1 above)
Comment (alyapunov) [reply changelogs/unreleased/gh-9994-next-prefix-iterator.md]: Fixed, thanks.

**Comment by alyapunov** [comment]
Both problems that lint was found are false-positive.

**Pull request merged**